### ISSUE TITLE
WCS WAT distortion with `wtype=tan` causes wcslib parser error

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#4089]
I get a `MemoryAllocationError` when trying to create a WCS object from this header:

```
SIMPLE  =                    T / Fits standard
BITPIX  =                  -32 / Bits per pixel
NAXIS   =                    2 / Number of axes
NAXIS1  =                 1024 / Axis length
NAXIS2  =                 1024 / Axis length
EXTEND  =                    T / File may contain extensions
ORIGIN  = 'NOAO-IRAF FITS Image Kernel July 2003' / FITS file originator
IRAF-TLM= '19:37:34 (05/10/2006)' / Time of last modification
OBJECT  = 'w51_hj_k'           / Name of the object observed
DATE    = '2006-10-05T22:37:34' / Date this file was written
EXPTIME =              20.0000          / Total integration time
MJD-OBS =       52842.20221769          / MJD start (2003-07-22T04:51:11.608)
DATE-OBS= '2003-07-22T04:51:11.6080'    / Observing date
ORIGFILE= 'NACO_0325.fits'              / Original File Name
INSTRUME= 'NAOS+CONICA'                 / Instrument used
TELESCOP= 'ESO-VLT-U4'                  / ESO Telescope Name
RA      =           290.917360          / 19:23:40.1 RA (J2000) pointing
DEC     =             14.51946          / 14:31:10.0 DEC (J2000) pointing
EQUINOX =                2000.          / Standard FK5
RADECSYS= 'FK5     '                    / FK5
LST     =            72459.260          / 20:07:39.260 LST
UTC     =            17468.000          / 04:51:08.000 UTC
OBSERVER= 'UNKNOWN '                    / Name of observer
PI-COI  = 'UNKNOWN '                    / Name(s) of proposer(s)
ALARM   = '        '                    / Active alarm(s), if any.
AIRMASS =              1.31600          / Averaged air mass
CRVAL1  =    290.9171447528058 / Coordinate at reference pixel in <axis
CRVAL2  =    14.51930969275231 / Coordinate at reference pixel in <axis
CRPIX1  =     478.319227457047 / Ref pixel in <axis direction>
CRPIX2  =     588.740275621415 / Ref pixel in <axis direction>
CDELT1  =  -7.5277800000000E-6 / Increment in <axis direction>
CDELT2  =  7.52778000000000E-6 / Increment in <axis direction>
CTYPE1  = 'RA---TAN'                    / Coordinate system of <axis direction>
CTYPE2  = 'DEC--TAN'                    / Coordinate system of <axis direction>
CD1_1   =         -7.52778E-06          / Translation matrix element
CD2_2   =          7.52778E-06          / Translation matrix element
ARCFILE = 'NACO.2003-07-22T04:51:11.608.fits' / Archive File Name
UT      = '04:51:08.000'                / UT at start
ST      = '20:07:39.260'                / ST at start
IMAGETYP= 'OBJECT  '                    / Observation type
WCSDIM  =                    2
LTV1    =     9.31922745704651
LTV2    =     2.74027562141418
LTM1_1  =                   1.
LTM2_2  =                   1.
WAT0_001= 'system=image'
WAT1_001= 'wtype=tan axtype=ra'
WAT2_001= 'wtype=tan axtype=dec'
NCOMBINE=                   31
OVERLAP = '[335:684,329:681]'
CRVAL1A =            290.91736
CRVAL2A =             14.51946
COMMENT   487 blank lines
END
```
on astropy 1.1.dev13127

@mdboom - could this be related to the precision improvements in wcs 5.9?

**Comment by saimn** [comment]
I get the same error with one file (I can paste the header if needed). With bisect I found that it comes from wcslib 5.8 (#3927).

```
$ ipython -c "from astropy import io, wcs; wcs.WCS(io.fits.getheader('data/obj/a370II.fits'))"
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-1-33c72c714720> in <module>()
----> 1 from astropy import io, wcs; wcs.WCS(io.fits.getheader('data/obj/a370II.fits'))

/home/conseil/lib/astropy/astropy/wcs/wcs.pyc in __init__(self, header, fobj, key, minerr, relax, naxis, keysel, colsel, fix, translate_units, _do_set)
    432                 wcsprm = _wcs.Wcsprm(header=header_bytes, key=key,
    433                                      relax=relax, keysel=keysel_flags,
--> 434                                      colsel=colsel)
    435             except _wcs.NoWcsKeywordsFoundError:
    436                 # The header may have SIP or distortions, but no core

MemoryError: Memory allocation error.
```

**Comment by mdboom** [comment]
Sorry -- looking at this has fallen to the wayside.  Looking at it now.

**Comment by mdboom** [comment]
The wcslib header parser throws parsing errors as of wcslib 5.x, and these weren't being handled correctly by astropy since previously the parser only ever through out-of-memory errors, so we just assumed it was one.  This is fixed in #4142.

Behind that we see that the error is caused by this keyword:

```
WAT1_001= 'wtype=tan axtype=ra'
```

`wcslib` handles a `wtype` value of `tnx` or `zpx` and considers anything else an error.  I'm getting a bit out of my depth on my FITS WCS standard knowledge about this one.  I think I'd prefer it if an unknown value causes the whole distortion to be ignored and it would move on.  This would at least replicate the behavior prior to wcslib 5.x when this keyword wasn't read at all.  I'll bring this to Mark Calabretta's attention to see what he thinks (and I'll rename this issue accordingly).

@saimn: Without seeing your header, I can't say whether your failure is due to the same reason or not.  Maybe open a new issue with your header and I'll have a look?

**Comment by saimn** [comment]
@mdboom : here is the header: https://gist.github.com/saimn/078444434d1a4cd76874
```
WAT0_001= 'system=image'                                                        
WAT1_001= 'wtype=tan axtype=ra'                                                 
WAT2_001= 'wtype=tan axtype=dec'  
```
So looks like the same issue. With your PR I get the "ValueError: Internal error in wcslib header parser" exception.

**Comment by saimn** [comment]
>  I'll bring this to Mark Calabretta's attention to see what he thinks (and I'll rename this issue accordingly).

Hi @mdboom ,
I know you may be busy with other things now, but are there any updates on this ? It would be annoying in astropy 1.1 was unable to read these files as before. Is there any workaround that could ship in astropy waiting for a better solution in wcslib (maybe) ? 
I found that `wcsbth` should be more permissive than `wcspih`, and I can already use this in astropy with the `keysel` keyword. This allows to get a WCS object without errors but I get a lot of warnings, the result seems correct but maybe there is a trap: 
```
In [3]: w = wcs.WCS(io.fits.getheader('data/obj/a370II.fits'), keysel=['image'])
WARNING: FITSFixedWarning: CTYPE1 = 'RA---TAN' /axis1 is in pixels 
image-header keyword CTYPEia in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CTYPE2 = 'DEC--TAN' /axis2 is in pixels 
image-header keyword CTYPEia in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: EPOCH = 0.0 /equinox of coordinates 
image-header keyword EPOCH in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: RADECSYS= 'FK5 ' 
the RADECSYS keyword is deprecated, use RADESYSa. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: EQUINOX = 2000. 
image-header keyword EQUINOXa in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CRVAL1 = 39.9720507246377 
image-header keyword CRVALia in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CRVAL2 = -1.58100543478261 
image-header keyword CRVALia in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CRPIX1 = 939.233218746011 
image-header keyword CRPIXja in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CRPIX2 = 759.829593560993 
image-header keyword CRPIXja in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CD1_1 = 2.30709933898762E-5 
image-header keyword CDi_ja in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CD1_2 = -5.2187244967449E-5 
image-header keyword CDi_ja in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CD2_1 = -5.2295565743288E-5 
image-header keyword CDi_ja in binary table. [astropy.wcs.wcs]
WARNING: FITSFixedWarning: CD2_2 = -2.3161588462530E-5 
image-header keyword CDi_ja in binary table. [astropy.wcs.wcs]

In [4]: w
Out[4]: 
WCS Keywords

Number of WCS axes: 2
CTYPE : 'RA---TAN'  'DEC--TAN'  
CRVAL : 39.972050724637697  -1.5810054347826099  
CRPIX : 939.23321874601095  759.82959356099298  
CD1_1 CD1_2  : 2.3070993389876198e-05  -5.2187244967449002e-05  
CD2_1 CD2_2  : -5.2295565743287998e-05  -2.3161588462530001e-05  
NAXIS    : 1909 1797

```

**Comment by saimn** [comment]
Can be closed, was fixed by #4239
